plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -419,37 +423,166 @@ public void downloadPlugins(List<Plugin> plugins) {
+            } else {
+                allPlugins.get(dependencyName).add(entry.getValue());
+            }
+        }
+    }
+
+    /**
+     * If there are any duplicated plugins in the list of all optimized plugins, removes them and their dependencies
+     *
+     * @param childDependencies map of all optimized dependencies, including duplicates
+     * @return optimized dependencies, which is a map of plugin name, plugin pairs
+     */
+    public Map<String, Plugin> optimize(Map<String, List<Plugin>> childDependencies) {
+        List<Plugin> duplicates = findDuplicates(childDependencies);
+
+        for (Plugin duplicate : duplicates) {
+            removeFromOptimalDependencies(duplicate, childDependencies);
+            List<Plugin> removedPluginsDependencies = new ArrayList<>(duplicate.getRecursiveDependencies().values());
+            for (Plugin dependency : removedPluginsDependencies) {
+                removeFromOptimalDependencies(dependency, childDependencies);
+            }
+        }
+
+        Map<String, Plugin> optimizedDependencies = new HashMap<>();
+
+        //after duplicates have been removed, should only be one plugin in the list
+        for (Map.Entry<String, List<Plugin>> entry : childDependencies.entrySet()) {
+            if (!entry.getValue().isEmpty()) {
+                optimizedDependencies.put(entry.getKey(), entry.getValue().get(0));
+            }
+        }
+        return optimizedDependencies;
+    }
+
+    /**
+     * Removes a plugin from the map corresponding to all plugin names and duplicated plugins
+     *
+     * @param pluginToRemove    plugin that will be removed
+     * @param childDependencies list of all of the optimized plugin dependencies of the children of the plugin to find
+     *                          optimal dependencies for
+     */
+    public void removeFromOptimalDependencies(Plugin pluginToRemove, Map<String, List<Plugin>> childDependencies) {
+        String pluginName = pluginToRemove.getName();
+        List<Plugin> plugins = childDependencies.get(pluginName);
+        Iterator<Plugin> pluginIterator = plugins.iterator();
+        while (pluginIterator.hasNext()) {
+            Plugin pluginInList = pluginIterator.next();
+            if (pluginInList == pluginToRemove) {
+                pluginIterator.remove();
+            }
+        }
+    }
+
+    /**
+     * Given a map of plugin names, and the list of plugins corresponding to that name, finds and returns the list of
+     * duplicate plugins
+     *
+     * @param childDependencies the set of all previously optimized dependencies and duplicates that resulted from
+     *                          combining these; this is a map of plugin name and a list of all plugins that share that
+     *                          name
+     * @return list of all duplicated plugins
+     */
+    public List<Plugin> findDuplicates(Map<String, List<Plugin>> childDependencies) {
```

Comment:
Since `Plugin` has a `compareTo` and a `equals` method that should do this type of replacement, have you looked at using Apache Collections [CollectionsUtils.intersection](https://commons.apache.org/proper/commons-collections/javadocs/api-4.4/org/apache/commons/collections4/CollectionUtils.html#intersection-java.lang.Iterable-java.lang.Iterable-) ?  This will return all the items that are in both lists without you having to do the comparison calculations yourself.

Reply:
Hmmm, will have to think about it. I defined plugin equivalency as having both the same name and same version. What I'm actually interested in this case is plugins with the same name, but different versions.

Reply:
You can also use custom comparators when needed :)